File: cpp/148.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>
#include <utility>
#include <map>
#include <set>
#include <string>
#include <deque>
#include <unordered_map>
#include <unordered_set>
#include <climits>
#include <numeric>
using namespace std;

template<typename T>
void prints(vector<T> &v){
  cout << "[";
  for(uint i=0; i<v.size()-1; ++i)
    cout << v[i] << ", ";
  if(v.size()) cout << v[v.size()-1];
  cout << "]\n";
}

struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
  ListNode *sortList(ListNode *head) {
    if(head == nullptr || head->next == nullptr)
      return head;
    ListNode h(0), *a, *b, *p;
    h.next = head;
    a = b = &h;
    while(b->next && b->next->next){
      a = a->next; b = b->next->next;
    }
    b = a->next;
    a->next = nullptr;
    a = head;
    a = sortList(a); b = sortList(b);
    h.next = nullptr; p = &h;
    while(a && b){
      if(a->val <= b->val){
        p->next = a; p = a; a = a->next;
      } else {
        p->next = b; p = b; b = b->next;
      }
    }
    p->next = a ? a : b;
    return h.next;
  }
};

int main(){
  auto sol = Solution();
  ListNode a(3), b(2), c(1);
  a.next = &b; b.next = &c;
  auto p = sol.sortList(&a);
  while(p)
    cout << p->val << endl, p = p->next;
}

File: python/148.py
class Solution(object):
    def sortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next:
            return head

        h = ListNode(None)
        p1 = p2 = h
        h.next = head
        while p2.next and p2.next.next:
            p2 = p2.next.next
            p1 = p1.next
        head2 = p1.next
        p1.next = None

        a, b = self.sortList(head), self.sortList(head2)
        h.next = None
        p = h
        while a and b:
            if a.val > b.val:
                a, b = b, a
            p.next = a
            p = a
            a = a.next
        p.next = a or b
        return h.next
